codersdk/workspacesdk/connector_internal_test.go

```diff
@@ -137,6 +142,138 @@ func TestTailnetAPIConnector_UplevelVersion(t *testing.T) {
+	agentID := uuid.UUID{0x55}
+	clientID := uuid.UUID{0x66}
+	fCoord := tailnettest.NewFakeCoordinator()
+	var coord tailnet.Coordinator = fCoord
+	coordPtr := atomic.Pointer[tailnet.Coordinator]{}
+	coordPtr.Store(&coord)
+	derpMapCh := make(chan *tailcfg.DERPMap)
+	defer close(derpMapCh)
+	eventCh := make(chan []*proto.TelemetryEvent, 1)
+	svc, err := tailnet.NewClientService(tailnet.ClientServiceOptions{
+		Logger:                 logger,
+		CoordPtr:               &coordPtr,
+		DERPMapUpdateFrequency: time.Millisecond,
+		DERPMapFn:              func() *tailcfg.DERPMap { return <-derpMapCh },
+		NetworkTelemetryHandler: func(batch []*proto.TelemetryEvent) {
+			eventCh <- batch
+		},
+	})
+	require.NoError(t, err)
+
+	svr := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		sws, err := websocket.Accept(w, r, nil)
+		if !assert.NoError(t, err) {
+			return
+		}
+		ctx, nc := codersdk.WebsocketNetConn(r.Context(), sws, websocket.MessageBinary)
+		err = svc.ServeConnV2(ctx, nc, tailnet.StreamID{
+			Name: "client",
+			ID:   clientID,
+			Auth: tailnet.ClientCoordinateeAuth{AgentID: agentID},
+		})
+		assert.NoError(t, err)
+	}))
+
+	fConn := newFakeTailnetConn()
+
+	uut := newTailnetAPIConnector(ctx, logger, agentID, svr.URL, &websocket.DialOptions{})
+	uut.runConnector(fConn)
+	require.Eventually(t, func() bool {
+		uut.clientMu.Lock()
+		defer uut.clientMu.Unlock()
+		return uut.client != nil
+	}, testutil.WaitShort, testutil.IntervalFast)
+
+	uut.SendTelemetryEvent(&proto.TelemetryEvent{
+		Id: []byte("test event"),
+	})
+
+	testEvents := testutil.RequireRecvCtx(ctx, t, eventCh)
+
+	require.Len(t, testEvents, 1)
+	require.Equal(t, []byte("test event"), testEvents[0].Id)
+}
+
+func TestTailnetAPIConnector_TelemetryUnimplemented(t *testing.T) {
+	t.Parallel()
+	ctx := testutil.Context(t, testutil.WaitShort)
+	logger := slogtest.Make(t, nil).Leveled(slog.LevelDebug)
+	agentID := uuid.UUID{0x55}
+	fConn := newFakeTailnetConn()
+
+	fakeDRPCClient := newFakeDRPCClient()
+	uut := &tailnetAPIConnector{
+		ctx:           ctx,
+		logger:        logger,
+		agentID:       agentID,
+		coordinateURL: "",
+		dialOptions:   &websocket.DialOptions{},
+		conn:          nil,
+		connected:     make(chan error, 1),
+		closed:        make(chan struct{}),
+		customDialFn: func() (proto.DRPCTailnetClient, error) {
+			return fakeDRPCClient, nil
+		},
+	}
+	uut.runConnector(fConn)
+	require.Eventually(t, func() bool {
+		uut.clientMu.Lock()
+		defer uut.clientMu.Unlock()
+		return uut.client != nil
+	}, testutil.WaitShort, testutil.IntervalFast)
+
+	fakeDRPCClient.telemeteryErorr = drpcerr.WithCode(xerrors.New("Unimplemented"), 0)
+	uut.SendTelemetryEvent(&proto.TelemetryEvent{})
+	require.False(t, uut.telemetryUnavailable.Load())
+
+	fakeDRPCClient.telemeteryErorr = drpcerr.WithCode(xerrors.New("Unimplemented"), drpcerr.Unimplemented)
+	uut.SendTelemetryEvent(&proto.TelemetryEvent{})
```

Comment:
It would also be good to verify that when `telemetryUnavailable` is true, `SendTelemetryEvent` no-ops. You could do that by adding a `sendTelemetryCalls int64` to `fakeDRPCClient`, accessing with atomic.Int64

Reply:
Is there any particular reason you would want to store an `int64` and access it using `atomic` functions instead of just storing an `atomic.Int64`?

Reply:
I said both but I meant either. atomic.Int64 is new